vault/raft.go

```diff
@@ -524,122 +524,232 @@ func (c *Core) raftSnapshotRestoreCallback(grabLock bool, sealNode bool) func(co
+			if err != nil {
+				return errwrap.Wrapf("error decoding raft bootstrap challenge: {{err}}", err)
+			}
+
+			eBlob := &physical.EncryptedBlobInfo{}
+			if err := proto.Unmarshal(challengeRaw, eBlob); err != nil {
+				return errwrap.Wrapf("error decoding raft bootstrap challenge: {{err}}", err)
+			}
+			raftInfo := &raftInformation{
+				challenge:           eBlob,
+				leaderClient:        apiClient,
+				leaderBarrierConfig: &sealConfig,
+				nonVoter:            nonVoter,
+			}
+
+			if c.seal.BarrierType() == seal.Shamir {
+				c.raftInfo = raftInfo
+				if err := c.seal.SetBarrierConfig(ctx, &sealConfig); err != nil {
+					return err
+				}
+
+				if !retry {
+					return nil
+				}
+
+				// Wait until unseal keys are supplied
+				c.logger.Info("wait until unseal keys are supplied")
+				c.raftInfo.joinInProgress = true
+				if atomic.LoadUint32(c.postUnsealStarted) != 1 {
+					return errors.New("waiting for unseal keys to be supplied")
+				}
+			}
+
+			if err := c.joinRaftSendAnswer(ctx, c.seal.GetAccess(), raftInfo); err != nil {
+				return errwrap.Wrapf("failed to send answer to raft leader node: {{err}}", err)
+			}
+
+			if c.seal.BarrierType() == seal.Shamir {
+				// Reset the state
+				c.raftInfo = nil
+
+				// In case of Shamir unsealing, inform the unseal process that raft join is completed
+				close(c.raftJoinDoneCh)
+			}
+
+			c.logger.Info("successfully joined the raft cluster", "leader_addr", leaderInfo.LeaderAPIAddr)
 
-		eBlob := &physical.EncryptedBlobInfo{}
-		if err := proto.Unmarshal(challengeRaw, eBlob); err != nil {
-			return errwrap.Wrapf("error decoding challenge: {{err}}", err)
-		}
-		raftInfo := &raftInformation{
-			challenge:           eBlob,
-			leaderClient:        apiClient,
-			leaderBarrierConfig: &sealConfig,
-			nonVoter:            nonVoter,
-		}
-		if c.seal.BarrierType() == seal.Shamir {
-			c.raftInfo = raftInfo
-			c.seal.SetBarrierConfig(ctx, &sealConfig)
 			return nil
 		}
 
-		if err := c.joinRaftSendAnswer(ctx, c.seal.GetAccess(), raftInfo); err != nil {
-			return errwrap.Wrapf("failed to send answer to leader node: {{err}}", err)
+		// Each join try goes through all the possible leader nodes and attempts to join
+		// them, until one of the attempt succeeds.
+		for _, leaderInfo := range leaderInfos {
+			err = joinLeader(leaderInfo)
+			if err == nil {
+				return nil
+			}
+			c.logger.Info("join attempt failed", "error", err)
 		}
 
-		return nil
+		return errors.New("failed to join any raft leader node")
 	}
 
-	switch retry {
+	switch leaderInfos[0].Retry {
 	case true:
 		go func() {
 			for {
-				// TODO add a way to shut this down
-				err := join()
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				err := join(true)
 				if err == nil {
 					return
 				}
-				c.logger.Error("failed to join raft cluster", "error", err)
-				time.Sleep(time.Second * 2)
+				c.logger.Error("failed to retry join raft cluster", "error", err, "retry", "5s")
```

Comment:
The error in this log line is redundant since it will always be "failed to join any raft leader node", let's just remove it.

```suggestion
				c.logger.Error("failed to retry join raft cluster", "retry", "2s")
```

Reply:
Done.